### PR TITLE
Add block io in the container stats of CRI #108575

### PR DESCRIPTION
**Pull request by yanghesong**
#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
Block I/O is the stats of tracking how many bytes are read/written from/to the block device on the node, for a particular container.
Refer to 
#108575
https://github.com/kubernetes-sigs/cri-tools/issues/898
#### Which issue(s) this PR fixes:
#108575

#### Does this PR introduce a user-facing change?

```release-note
Block io is supported in CRI. Block io state can be aquire by CRI.
```
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by pacoxu** [comment]
> @yanghesong: Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository.
> Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).

Please follow the guide above to rebase the branch.

**Comment by yanghesong** [comment]
> > @yanghesong: Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository.
> > Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).
> 
> Please follow the guide above to rebase the branch.

Done.

**Comment by pacoxu** [comment]
/triage accepted
/priority important-longterm

since containerd and cri-o block io support is almost ready.

> Current status of blockio throttling in container runtimes:
> 
> CRI-O: support released in v1.22, see https://github.com/cri-o/cri-o/pull/4873.
> Containerd: a fully compatible https://github.com/containerd/containerd/pull/5490 is open.

**Comment by pacoxu** [comment]
/cc @qiutongs 

For `docker stats`, https://docs.docker.com/engine/reference/commandline/stats/ 


Column name | Description
-- | --
BLOCK I/O | The amount of data the container has read to and written from block devices on the host

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 1]
```diff
@@ -1523,6 +1523,8 @@ message ContainerStats {
     MemoryUsage memory = 3;
     // Usage of the writable layer.
     FilesystemUsage writable_layer = 4;
+    // Usage of the Block IO
+    BlockIOUsage block_io_usage = 5;
```
Comment (mikebrow) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1527]:
```suggestion
    BlockIOUsage block_io = 5;
```
noting above cpu and memory field name pattern..

Comment (mikebrow) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1527]: same for v1alpha proto

**Review by mikebrow — COMMENTED** [review]
LGTM with nit on field naming pattern

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (yanghesong) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]: Thank you for your suggestion. The modification has been committed.

**Comment by pacoxu** [comment]
/test pull-kubernetes-unit
for a flake

**Review by pacoxu — COMMENTED** [review]
/lgtm

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by yanghesong** [comment]
> If you still need this PR then please rebase, if not, please close the PR

Rebase has been done on Nov 9.

**Comment by dims** [comment]
> Rebase has been done on Nov 9.

thanks!

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (ruiwen-zhao) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1611]: I am still seeing `block_io_usage` here...

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 2]
```diff
@@ -1607,6 +1607,8 @@ message ContainerStats {
...
```
Comment (ruiwen-zhao) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1611]: I am still seeing block_io_usage here...

**Comment by ruiwen-zhao** [comment]
Sorry for the late review. 

1. Please fix the name `block_io_usage ` as mentioned above. 
2. Can you make the two commits into one, since they are for the same purpose?

**Comment by yanghesong** [comment]
> Sorry for the late review.
> 
> 1. Please fix the name `block_io_usage ` as mentioned above.
> 2. Can you make the two commits into one, since they are for the same purpose?

Thank you for your reminder, the modification has been completed.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 2 above)
Comment (yanghesong) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]: Done.

**Comment by yanghesong** [comment]
> Sorry for the late review.
> 
> 1. Please fix the name `block_io_usage ` as mentioned above.
> 2. Can you make the two commits into one, since they are for the same purpose?

Thanks for the reminder, please help me to review again.

**Review by marquiz — COMMENTED** [review]
/lgtm

**Comment by marquiz** [comment]
ach @yanghesong please rebase

**Review by marquiz — COMMENTED** [review]
Thx @yanghesong 
/lgtm

**Comment by mrunalp** [comment]
@haircommander @bobbypage ptal

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 3]
```diff
@@ -1627,6 +1627,8 @@ message ContainerStats {
     MemoryUsage memory = 3;
     // Usage of the writable layer.
     FilesystemUsage writable_layer = 4;
+    // Usage of the Block IO
+    BlockIOUsage block_io = 5;
```
Comment (haircommander) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1631]: Is this possibly relevant to the windows stats object as well?

cc @marosset

**Comment by haircommander** [comment]
so this brings up an architectural question I don't have a great answer for yet: should the stats in this object only be for populating Kubelet's /stats/summary API or should they be extended for debugging with crictl? I am not sure, but I lean to the former (as that was the purpose of adding it in the first place). if that's the case--since kubelet doesn't use this in /stats/summary, I'm not sure we should add this metric

**Comment by pacoxu** [comment]
> CRI will be changed in version of v1 and v1alpha

The release note is not clear. You should mention that block io is supported.

**Comment by yanghesong** [comment]
> > CRI will be changed in version of v1 and v1alpha
> 
> The release note is not clear. You should mention that block io is supported.

Done. Thanks!

**Review by mikebrow — APPROVED** [review]
LGTM

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 3 above)
Comment (mikebrow) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1631]: were we going to deprecate the duplicate object definitions?

Comment (haircommander) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1631]: no the path forward is to have them defined for each OS to allow customized stats for each (when eventually appropriate)

**Comment by mikebrow** [comment]
> so this brings up an architectural question I don't have a great answer for yet: should the stats in this object only be for populating Kubelet's /stats/summary API or should they be extended for debugging with crictl? I am not sure, but I lean to the former (as that was the purpose of adding it in the first place). if that's the case--since kubelet doesn't use this in /stats/summary, I'm not sure we should add this metric

seems like a useful stat.. rancher/cloud client tools may also use it.. 

I'd expect /stats/summary to have a `--raw` option, at min, that would show the new field?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 3 above)
Comment (mikebrow) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1631]: thx :-) let it be written!

**Comment by pacoxu** [comment]
> PR needs rebase.

Do you have time to rebase this? We may have this in v1.29 release.

**Comment by yanghesong** [comment]
> > PR needs rebase.
> 
> Do you have time to rebase this? We may have this in v1.29 release.

I will rebase it as soon as possible.

**Comment by yanghesong** [comment]
> > PR needs rebase.
> 
> Do you have time to rebase this? We may have this in v1.29 release.

Rebase has been done.  /cc @pacoxu

**Comment by haircommander** [comment]
personally I am a NACK on this. I understand the motivation but I don't think this is in scope of /stats/summary API. I think this is better suited for a prometheus metric or something else. I am not staunchly opposed so I could be convinced but I think we should be wary of what we expose here
